### PR TITLE
Catalog ingestion: code refactoring

### PR DESCRIPTION
**Pull request by garmashdev**
# Description
- Refactored 2.2.8/2.3.0 support, added isModuleEnabled method instead di.xml class name insertion as string
- Changed in_stock/out_of_stock catalog ingestion triggering: now it's support multiple sales channels and multiple store sources assigned instead triggering each time when at least one of source item was changed.

Fixes: (link ticket)

#changelog Catalog ingestion: code refactoring

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 

- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my ticket link and provided a changelog message.

### CODE REVIEW

**Comment by garmashdev** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1590** <a href="https://app.graphite.dev/github/pr/BoltApp/bolt-magento2/1590" target="_blank">Graphite</a>  👈
    * **PR #1591** <a href="https://app.graphite.dev/github/pr/BoltApp/bolt-magento2/1591" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/BoltApp/bolt-magento2/1590?utm_source=stack-comment).

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (vitaliyreznikov) [inline Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:90]: Any specific reason why we want to call the plugin directly?
I think it's maybe cleaner if we copy its code to our helpers.

File: Model/CatalogIngestion/ProductEventManager.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventManager.php:148]: for first one: //create product have more priority than update
when we are creating a new product, the stock items update plugins is triggered earlier then product observer. In this case we have update event for not created product.

for second one: //delete product have more priority than create/update
if we create product and then remove we will have only one product event -> create which is incorrect.

File: Model/CatalogIngestion/ProductEventManager.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventManager.php:213]: not it is just additional check to prevent double sending request in plugins stack, related to catalog inventory & legacy magento inventory flows. When catalog inventory doesn't have any source items its working as legacy and triggers update 2 times one for source items second for native catalog inventory.

File: Model/CatalogIngestion/ProductEventPublisher.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventPublisher.php:38]: magento 2.3.0 have different default API topic names definitions.

File: Plugin/Magento/Inventory/Model/SourceItem/Command/DecrementSourceItemQtyPlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/DecrementSourceItemQtyPlugin.php:106]: not, we need to have salable status before update and after source items update.

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:104]: not, we need to have salable status before update and after source items update.

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:90]: we need correct values of product status before and after update, default indexer plugin can't be called in around plugin automatically, but reindex is required to have correct values of product stock status.

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:117]: updated.

File: Model/CatalogIngestion/ProductEventProcessor.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventProcessor.php:260]: oldStockItem can be null in case when we creating a new product. We can't skip this check for new products only.

File: Plugin/Magento/InventorySalesApi/Api/PlaceReservationsForSalesEventPlugin.php
Comment (garmashdev) [reply Plugin/Magento/InventorySalesApi/Api/PlaceReservationsForSalesEventPlugin.php:80]: updated

File: Model/CatalogIngestion/ProductEventPublisher.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventPublisher.php:100]: - comment added.
- system configuration comment updated.
- it requires additional plugins implementation, magento don't have default visible config xml declaration based on magento versions. Currently all options instead one : "catalog ingestion enabled/disabled" are hidden.

File: Model/CatalogIngestion/ProductEventProcessor.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventProcessor.php:157]: updated.

File: Model/CatalogIngestion/ProductEventPublisher.php
Comment (vitaliyreznikov) [reply Model/CatalogIngestion/ProductEventPublisher.php:38]: can we add links to the documentation here?

File: Model/CatalogIngestion/ProductEventManager.php
Comment (garmashdev) [reply Model/CatalogIngestion/ProductEventManager.php:148]: updated to elseif

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:106]: updated to before/after

File: Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php
Comment (garmashdev) [reply Plugin/Magento/Inventory/Model/SourceItem/Command/SourceItemSavePlugin.php:92]: updated to before/after

File: Model/CatalogIngestion/ProductEventProcessor.php
Comment (vitaliyreznikov) [reply Model/CatalogIngestion/ProductEventProcessor.php:261]: As discussed should be
`(!is_null($oldStockItem)) ? (bool)$oldStockItem->getIsInStock() : null,`

File: Model/CatalogIngestion/ProductEventPublisher.php
Comment (vitaliyreznikov) [reply Model/CatalogIngestion/ProductEventPublisher.php:102]: Sounds good! Thank you!

File: Plugin/Magento/InventorySalesApi/Api/PlaceReservationsForSalesEventPlugin.php
Comment (vitaliyreznikov) [reply Plugin/Magento/InventorySalesApi/Api/PlaceReservationsForSalesEventPlugin.php:87]: excellent!

**Review by vitaliyreznikov — APPROVED** [review]

**Comment by garmashdev** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/BoltApp/bolt-magento2/1590?utm_source=stack-comment).


Job Reference: HiIaPPQEAK5uY3radM0x

**Pull request merged**
